src/WebJobs.Script.WebHost/Management/InstanceManager.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Script.WebHost.Models;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost.Management
+{
+    public class InstanceManager : IInstanceManager
+    {
+        private static HostAssignmentContext _assignmentContext;
+        private static object _assignmentLock = new object();
+
+        private readonly ScriptHostManager _scriptHostManager;
+        private readonly WebHostSettings _webHostSettings;
+        private readonly ILogger _logger;
+
+        public InstanceManager(WebScriptHostManager scriptHostManager, WebHostSettings webHostSettings, ILoggerFactory loggerFactory)
+        {
+            _scriptHostManager = scriptHostManager;
+            _webHostSettings = webHostSettings;
+            _logger = loggerFactory.CreateLogger(nameof(InstanceManager));
+        }
+
+        public bool TryAssign(HostAssignmentContext assignmentContext)
+        {
+            if (_assignmentContext == null)
+            {
+                lock (_assignmentLock)
+                {
+                    if (_assignmentContext != null)
+                    {
+                        return _assignmentContext.Equals(assignmentContext);
+                    }
+                    _assignmentContext = assignmentContext;
+                }
+
+                // Queue the task in background, and return to caller immediately
+                Task.Run(() => Specialize(assignmentContext));
+
+                return true;
```

Comment:
Based on the behavior, it would be good to have a different name for this method (something like `ScheduleAssignment`, `StartAssignment`, etc.). Looking at the API `TryAssign` returning a `bool` makes it sound like the assignment was succesful if true is returned, which is not the case.

Reply:
makes sense.